File: scripts/database_integration_scripts_V2/legacy/62_create_works.py
"""62 — Create works table.

Mirrors `enhance_db/src/bin/62_create_works.rs`.

  Inputs : data/all_humans/all_human_works.tsv
              (header + columns: individual_id <TAB> work_id <TAB> relationship)
           data/all_humans/work_labels.json
              {work_qid: english_label}
           individuals (for the joined name_en)
  Output : works (id PK auto, individual_id, individual_name, work_id,
                    work_name, relationship)

Usage
-----
    python3 62_create_works.py
    python3 62_create_works.py --full
"""

from __future__ import annotations

import json
import sqlite3
import tempfile
from pathlib import Path

from tqdm import tqdm

from common import (
    ALL_HUMANS_DIR,
    insert_rows,
    load_json,
    log,
    open_db,
    parse_run_mode,
)

WORKS_TSV = ALL_HUMANS_DIR / "all_human_works.tsv"
LABELS_JSON = ALL_HUMANS_DIR / "work_labels.json"


def run(
    conn: sqlite3.Connection,
    tsv_path: Path = WORKS_TSV,
    labels_path: Path = LABELS_JSON,
) -> int:
    log("[DB] 62: Create works table...")
    work_labels = load_json(labels_path)
    log(f"[DB] {len(work_labels)} work labels")

    indiv_names: dict[str, str] = {}
    for qid, name in conn.execute("SELECT wikidata_id, name_en FROM individuals"):
        if name:
            indiv_names[qid] = name
    log(f"[DB] {len(indiv_names)} individual names loaded")

    conn.execute("DROP TABLE IF EXISTS works")
    conn.execute(
        """
        CREATE TABLE works (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            individual_id TEXT NOT NULL,
            individual_name TEXT,
            work_id TEXT NOT NULL,
            work_name TEXT,
            relationship TEXT NOT NULL
        )
        """
    )

    cur = conn.cursor()
    cur.execute("BEGIN")
    n = 0
    missing_indiv = 0
    missing_label = 0
    with open(tsv_path, "r", encoding="utf-8") as fh:
        fh.readline()  # header
        for line in tqdm(fh, desc="62", unit="line"):
            parts = line.rstrip("\n").split("\t")
            if len(parts) < 3:
                continue
            iid, wid, rel = parts[0], parts[1], parts[2]
            iname = indiv_names.get(iid)
            if iname is None:
                missing_indiv += 1
            wname = work_labels.get(wid)
            if wname is None:
                missing_label += 1
            cur.execute(
                "INSERT INTO works (individual_id, individual_name, work_id, "
                "work_name, relationship) VALUES (?,?,?,?,?)",
                (iid, iname, wid, wname, rel),
            )
            n += 1
            if n % 100_000 == 0:
                conn.commit()
                cur.execute("BEGIN")
    conn.commit()
    log(f"[DB] inserted {n} (missing indiv: {missing_indiv}, missing label: {missing_label})")

    for sql in (
        "CREATE INDEX IF NOT EXISTS idx_works_individual ON works(individual_id)",
        "CREATE INDEX IF NOT EXISTS idx_works_work ON works(work_id)",
        "CREATE INDEX IF NOT EXISTS idx_works_rel ON works(relationship)",
    ):
        conn.execute(sql)
    conn.commit()
    return n


def _sample_main() -> None:
    with tempfile.TemporaryDirectory() as tmp:
        tsv = Path(tmp) / "works.tsv"
        tsv.write_text(
            "individual_id\twork_id\trelationship\n"
            "Q1\tQ100\tP50\n"
            "Q1\tQ101\tP170\n"
            "Q2\tQ100\tP86\n"
        )
        labels_path = Path(tmp) / "labels.json"
        labels_path.write_text(json.dumps({
            "Q100": "Magnum Opus", "Q101": "Sonnet 1",
        }))
        db = Path(tmp) / "sample.sqlite3"
        with sqlite3.connect(db) as seed:
            seed.execute(
                "CREATE TABLE individuals (wikidata_id TEXT PRIMARY KEY, name_en TEXT)"
            )
            insert_rows(seed, "individuals", [
                {"wikidata_id": "Q1", "name_en": "Alice"},
                {"wikidata_id": "Q2", "name_en": "Bob"},
            ])
        with open_db(db) as conn:
            run(conn, tsv_path=tsv, labels_path=labels_path)
            for r in conn.execute(
                "SELECT individual_id, individual_name, work_id, work_name, "
                "relationship FROM works"
            ):
                log(f"  {r}")


if __name__ == "__main__":
    if parse_run_mode() == "full":
        with open_db() as conn:
            run(conn)
    else:
        _sample_main()

File: scripts/database_integration_scripts/enhance_db/src/bin/62_create_works.rs
/// Create works table from data/all_humans/all_human_works.tsv + work_labels.json.
///
/// Schema:
///   id              INTEGER PRIMARY KEY AUTOINCREMENT
///   individual_id   TEXT       (Wikidata QID of the human)
///   individual_name TEXT       (joined from individuals.name_en)
///   work_id         TEXT       (Wikidata QID of the work)
///   work_name       TEXT       (English label, joined from work_labels.json)
///   relationship    TEXT       (Wikidata property: P50, P170, P86, P57, P162, P98, P175, P110, P58)
///
/// One row per (individual, work, relationship) — a work counted under multiple
/// roles will appear multiple times (e.g. director + screenwriter).
use anyhow::{Context, Result};
use rusqlite::{params, Connection};
use std::collections::HashMap;
use std::fs::{self, File};
use std::io::{BufRead, BufReader, Write};
use std::time::Instant;

const DB_PATH: &str = "data/humans_clean.sqlite3";
const WORKS_TSV: &str = "data/all_humans/all_human_works.tsv";
const LABELS_JSON: &str = "data/all_humans/work_labels.json";
const TASK_LOG: &str = "task.log";

fn now_clock() -> String {
    let dur = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap();
    let secs = dur.as_secs();
    let h = (secs % 86400) / 3600;
    let m = (secs % 3600) / 60;
    let s = secs % 60;
    format!("{:02}:{:02}:{:02} UTC", h, m, s)
}

fn log(msg: &str) {
    let line = format!("[{}] {}", now_clock(), msg);
    println!("{}", line);
    let mut f = fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(TASK_LOG)
        .unwrap();
    writeln!(f, "{}", line).unwrap();
}

fn elapsed(start: Instant) -> String {
    let s = start.elapsed().as_secs();
    if s < 60 {
        format!("{}s", s)
    } else if s < 3600 {
        format!("{}m {}s", s / 60, s % 60)
    } else {
        format!("{}h {}m {}s", s / 3600, (s % 3600) / 60, s % 60)
    }
}

fn main() -> Result<()> {
    let total_start = Instant::now();
    log("=== Step 62: Create works table ===");

    // Load work labels
    let raw = fs::read_to_string(LABELS_JSON)
        .with_context(|| format!("reading {}", LABELS_JSON))?;
    let work_labels: HashMap<String, String> =
        serde_json::from_str(&raw).context("parsing work labels JSON")?;
    log(&format!("[62] Loaded {} work labels", work_labels.len()));

    let conn = Connection::open(DB_PATH)?;
    conn.execute_batch(
        "PRAGMA journal_mode=WAL; PRAGMA synchronous=NORMAL; PRAGMA cache_size=-2000000;",
    )?;

    // Load individual names (wikidata_id -> name_en)
    let mut indiv_names: HashMap<String, String> = HashMap::new();
    {
        let mut s = conn.prepare("SELECT wikidata_id, name_en FROM individuals")?;
        let rows = s.query_map([], |r| {
            let id: String = r.get(0)?;
            let name: Option<String> = r.get(1)?;
            Ok((id, name))
        })?;
        for row in rows {
            let (id, name) = row?;
            if let Some(n) = name {
                indiv_names.insert(id, n);
            }
        }
    }
    log(&format!("[62] Loaded {} individual names", indiv_names.len()));

    conn.execute_batch("DROP TABLE IF EXISTS works;")?;
    conn.execute_batch(
        "CREATE TABLE works (
            id              INTEGER PRIMARY KEY AUTOINCREMENT,
            individual_id   TEXT NOT NULL,
            individual_name TEXT,
            work_id         TEXT NOT NULL,
            work_name       TEXT,
            relationship    TEXT NOT NULL
        );",
    )?;
    log("[62] Created table works");

    let step = Instant::now();
    conn.execute_batch("BEGIN TRANSACTION;")?;

    let mut inserted: i64 = 0;
    let mut skipped_header = false;
    let mut missing_indiv: i64 = 0;
    let mut missing_label: i64 = 0;

    {
        let mut ins = conn.prepare(
            "INSERT INTO works (individual_id, individual_name, work_id, work_name, relationship)
             VALUES (?1, ?2, ?3, ?4, ?5)",
        )?;

        let f = File::open(WORKS_TSV).with_context(|| format!("opening {}", WORKS_TSV))?;
        let reader = BufReader::new(f);

        for line in reader.lines() {
            let line = line?;
            if !skipped_header {
                skipped_header = true;
                continue;
            }
            let parts: Vec<&str> = line.split('\t').collect();
            if parts.len() < 3 {
                continue;
            }
            let individual_id = parts[0];
            let work_id = parts[1];
            let relationship = parts[2];

            let individual_name = indiv_names.get(individual_id);
            if individual_name.is_none() {
                missing_indiv += 1;
            }
            let work_name = work_labels.get(work_id);
            if work_name.is_none() {
                missing_label += 1;
            }

            ins.execute(params![
                individual_id,
                individual_name,
                work_id,
                work_name,
                relationship,
            ])?;
            inserted += 1;

            if inserted % 250_000 == 0 {
                log(&format!(
                    "[62]   Inserted {} rows ({})",
                    inserted,
                    elapsed(step)
                ));
            }
        }
    }

    conn.execute_batch("COMMIT;")?;
    log(&format!(
        "[62] Insert complete: {} rows in {}",
        inserted,
        elapsed(step)
    ));

    let idx = Instant::now();
    conn.execute_batch(
        "CREATE INDEX IF NOT EXISTS idx_works_individual ON works(individual_id);
         CREATE INDEX IF NOT EXISTS idx_works_work       ON works(work_id);
         CREATE INDEX IF NOT EXISTS idx_works_rel        ON works(relationship);",
    )?;
    log(&format!("[62] Indexes created ({})", elapsed(idx)));

    let total_rows: i64 = conn.query_row("SELECT COUNT(*) FROM works", [], |r| r.get(0))?;
    let unique_humans: i64 = conn.query_row(
        "SELECT COUNT(DISTINCT individual_id) FROM works",
        [],
        |r| r.get(0),
    )?;
    let unique_works: i64 = conn.query_row(
        "SELECT COUNT(DISTINCT work_id) FROM works",
        [],
        |r| r.get(0),
    )?;

    log("[62] === Summary ===");
    log(&format!("[62]   Rows:                 {}", total_rows));
    log(&format!("[62]   Distinct individuals: {}", unique_humans));
    log(&format!("[62]   Distinct works:       {}", unique_works));
    log(&format!("[62]   Missing indiv name:   {}", missing_indiv));
    log(&format!("[62]   Missing work label:   {}", missing_label));

    log("[62] Per-relationship counts:");
    let mut s = conn.prepare(
        "SELECT relationship, COUNT(*) FROM works GROUP BY relationship ORDER BY 2 DESC",
    )?;
    let rows: Vec<(String, i64)> = s
        .query_map([], |r| Ok((r.get(0)?, r.get(1)?)))?
        .filter_map(|r| r.ok())
        .collect();
    for (rel, n) in rows {
        log(&format!("[62]   {:<5} {:>12}", rel, n));
    }

    log("[62] Sample rows:");
    let mut s = conn.prepare(
        "SELECT individual_id, individual_name, work_id, work_name, relationship
         FROM works LIMIT 5",
    )?;
    let rows: Vec<(String, Option<String>, String, Option<String>, String)> = s
        .query_map([], |r| {
            Ok((r.get(0)?, r.get(1)?, r.get(2)?, r.get(3)?, r.get(4)?))
        })?
        .filter_map(|r| r.ok())
        .collect();
    for (iid, iname, wid, wname, rel) in rows {
        log(&format!(
            "[62]   {} ({:?}) -> {} ({:?}) [{}]",
            iid, iname, wid, wname, rel
        ));
    }

    log(&format!(
        "=== Step 62 complete (total: {}) ===",
        elapsed(total_start)
    ));
    Ok(())
}
